manuals/pffinfo.1
.Dd July 29, 2012
.Dt pffinfo
.Os libpff
.Sh NAME
.Nm pffinfo
.Nd determines information about a Personal Folder File (OST, PAB and PST)
.Sh SYNOPSIS
.Nm pffinfo
.Op Fl c Ar codepage
.Op Fl ahvV
.Va Ar source
.Sh DESCRIPTION
.Nm pffinfo
is a utility to determine information about a Personal Folder File (OST, PAB and PST)
.Pp
.Nm pffinfo
is part of the
.Nm libpff
package.
.Nm libpff
is a library to support the Personal Folder File (OST, PAB and PST) format
.Pp
.Ar source
is the source file.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl a
shows allocation information
.It Fl c Ar codepage
specify the codepage of ASCII strings, options: ascii, windows-1250, windows-1251, windows-1252 (default), windows-1253, windows-1254, windows-1255, windows-1256, windows-1257 or windows-1258
.It Fl h
shows this help
.It Fl v
verbose output to stderr
.It Fl V
print version
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# pffinfo -a file.pst
pffinfo 20110107

Personal Folder File information:
	File size:		271360 bytes
	File content type:	Personal Storage Tables (PST)
	File type:		64-bit
	Encryption type:	compressible

Message store:
	Folders:		Subtree, Wastbox, Finder
	Password checksum:	N/A

Unallocated index node blocks:
	00018944 (0x00004a00) - 00020480 (0x00005000) size: 1536
.Dl        ...
	01051648 (0x00100c00) - 01052160 (0x00100e00) size: 512

Unallocated data blocks:
	00025024 (0x000061c0) - 00025536 (0x000063c0) size: 512
.Dl        ...
	01353664 (0x0014a7c0) - 01541120 (0x00178400) size: 187456

.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh BUGS
Please report bugs of any kind to <joachim.metz@gmail.com> or on the project website:
http://code.google.com/p/libpff/
.Sh AUTHOR
These man pages were written by Joachim Metz.
.Sh COPYRIGHT
Copyright 2008-2012 Joachim Metz <joachim.metz@gmail.com>.
This is free software; see the source for copying conditions. There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.Sh SEE ALSO
.Xr pffexport 1


ChangeLog
TODO
* pffexport/export handle:
  - handle MIME attachments
  - protect against duplicate name clashes
  - dump attachment item values
  - add support for LIBPFF_ITEM_TYPE_COMMON (IPM)
  - separate item file into separate structure
  - move message specific funtionality into message file ?
* item file:
  - create function to write enum values to the item file
  - create function to write flag values to the item file
* pypff
  - add Py_BEGIN_ALLOW_THREADS and Py_END_ALLOW_THREADS
  - add attributes
  - improved pydoc comment
* code clean up
  - what's the status of table block ?
  - unused and goto
* libpff: UTF8/UTF16
  - libpff_message_get_plain_text_body_size
  - libpff_message_get_rtf_body
* update codegear files
  - re-enable them

* what about debug export and error notifications

* code clean up
  - refactor table index into fdata list?
  - does table index contain full copy of table data ? if so refactor
* add table API ?
  - read data on demand like libesedb
  - move low-level item functions to table API
  - should table manage local descriptors and cache ?
  - separate item values code from table ?

* remove ascii codepage param from libpff_item_values_get_entry_value ?

* add const to export path in export handle
* replace internal_file by io_handle in item

* add item (descriptor identifier) type support
  - e.g. do not read a table if item does not contain a table

* cache local descriptor node (list element) in local descriptor value ?
  - reduces search for list element

libpff
* rename value (type) to table entry value (type)
* check node values ranges in index tree
* add abort support in long running function
  - make sure to return library to safe state
  - make sure to reset library abort value
* reduce impact of duplicate MAPI entry types
* validate table entry identifiers of level >= 1 with those of lower levels?
* codepage support
  - implement libuna-libpff codepage mapping function

libpff - check/test
* check support for multiple level 8c tables

libpff - API
* message store
  - add functions to retrieve special folder entry items
* add function to retrieve:
  - the file offset ranges used by an item ?
* name to id map
  - API access to entire name to id map ?
  - add libpff_file_get_name_to_id_map to include file?
* add a function to retrieve the total number of items?
  - maintain the total number of items?
  - maintain the total number of folder items?
* item
  - add read in buffers functions for entry values like message body?

libpff - error tollerance and recovery
* handle multiple descriptors in libpff_io_handle_get_descriptor_index_value_by_identifier
  and libpff_io_handle_get_offset_index_value_by_identifier
* analyze remaining/deleted content
* fragment recovery
  - add fragment support
  - allow fragments/partial items (add upper level reporting)
  - allow items without/with invalid local descriptors?
* item recovery
  - improve local descriptors read test - to read and check entire tree
  - improve performance of deleted offset index value duplicate check
  - handle duplicate offset index value identifiers
  - check fragment scan for correct looping of buffer
  - group recovered items with the same parent (do not link with allocated parent)
  - add support for recovered embedded items
* CRC error recovery
  - detect CRC32 corruption
  - detect data corruption
* automatically detect if allocation tables are useable ?

libpff - performance
* allocation tables
  - read allocation tables on demand ?
  - for large list support replace allocation offset list by tree variant
* optimized version of libpff_encryption_decrypt ?
* item hierarchy (tree)
  - build on demand ?
  - flag if sub nodes have been created ?
* item values
  - read directly into external buffer?
* attachment data
  - read directly into external buffer?
* does pre-reading item values of attachment improve performance ?

libpff - file format support
* add free maps support
* add support for cell existence
* add support for cc table type
* add support for the density list

libpff - clean up and refactoring
* libpff_table
  - cleaner reusable data structure? (libfdata)
  - create 7c table read entry definitions function
  - create ac table read entry definitions function
* have multi type work with data reference (in libpff_item) ?
* tree type implement use of initialize function (grep sizeof( libpff_tree_node_t ) )
* refactor direct use of array, list, and tree value attributes
* improve force decryption check

pfftools
* pffinfo
  - add byte size string
* pffexport
  - add flag to overwrite export directory?
  - export the data in the pff file (embedded objects)
  - use identifier numbering for traceability?
  - add support for export to mbox, eml (ole2)
  - support control code in subject
  - provide export summary
  - allow to set date format
  - remove ftk output
  - create outlook html output
  - add support for setting recovery method
* export handle
  - try to export items without item type ?
  - handle duplicate attachment names (add suffix number)
  - export attachment to separate subdirectory
  - improve filename sanitation - do a sanity test
  - use close related character replacements to improve readability
  - check into unable to export attachment not being notified in export log
  - correctly handle skipping an item without class name
  - check if 0x0e1f (PidTagRtfInSync) is set and export alternating RTF body
  - move target path to common functions as much as possible
  - handle multi message body formats for non email items
    add function to handle prefered and available message body
    change text message body export of item types into prefered/available message body
  - item support
    - IPM.DistList
    - IPM.FAX
    - IPM.Post
    - IPM.StickyNote
    - IPM.Configuration.RssRule
    - IPM.Configuration.HomeTimeZone
    - IPM.Configuration.AvailabilityOptions
    - IPM.Configuration.Calendar
    - IPM.Configuration.CategoryList
    - IPM.Configuration.WorkHours

pfftools/pffinfo.c
/*
 * Shows information obtained from a Personal Folder File (OST, PAB and PST)
 *
 * Copyright (c) 2008-2012, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This software is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with this software.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <memory.h>
#include <types.h>

#include <stdio.h>

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#include "info_handle.h"
#include "pffinput.h"
#include "pffoutput.h"
#include "pfftools_libcerror.h"
#include "pfftools_libclocale.h"
#include "pfftools_libcnotify.h"
#include "pfftools_libcstring.h"
#include "pfftools_libcsystem.h"
#include "pfftools_libpff.h"

info_handle_t *pffinfo_info_handle = NULL;
int pffinfo_abort                  = 0;

/* Prints the executable usage information
 */
void usage_fprint(
      FILE *stream )
{
	if( stream == NULL )
	{
		return;
	}
	fprintf( stream, "Use pffinfo to determine information about a Personal Folder File (OST, PAB\n"
	                 "and PST).\n\n" );

	fprintf( stream, "Usage: pffinfo [ -c codepage ] [ -ahvV ] source\n\n" );

	fprintf( stream, "\tsource: the source file\n\n" );

	fprintf( stream, "\t-a:     shows allocation information\n" );
	fprintf( stream, "\t-c:     codepage of ASCII strings, options: ascii, windows-1250,\n"
	                 "\t        windows-1251, windows-1252 (default), windows-1253,\n"
	                 "\t        windows-1254, windows-1255, windows-1256, windows-1257\n"
	                 "\t        or windows-1258\n" );
	fprintf( stream, "\t-h:     shows this help\n" );
	fprintf( stream, "\t-v:     verbose output to stderr\n" );
	fprintf( stream, "\t-V:     print version\n" );
}

/* Signal handler for pffinfo
 */
void pffinfo_signal_handler(
      libcsystem_signal_t signal LIBCSYSTEM_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function   = "pffinfo_signal_handler";

	LIBCSYSTEM_UNREFERENCED_PARAMETER( signal )

	pffinfo_abort = 1;

	if( pffinfo_info_handle != NULL )
	{
		if( info_handle_signal_abort(
		     pffinfo_info_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal info handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
	if( libcsystem_file_io_close(
	     0 ) != 0 )
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( LIBCSTRING_HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	libcerror_error_t *error                              = NULL;
	libcstring_system_character_t *option_ascii_codepage = NULL;
	libcstring_system_character_t *source                = NULL;
	char *program                                        = "pffinfo";
	libcstring_system_integer_t option                   = 0;
	uint8_t show_allocation_information                  = 0;
	int result                                           = 0;
	int verbose                                          = 0;

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "pfftools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( libcsystem_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize system values.\n" );

		goto on_error;
	}
	pffoutput_version_fprint(
	 stdout,
	 program );

	while( ( option = libcsystem_getopt(
	                   argc,
	                   argv,
	                   _LIBCSTRING_SYSTEM_STRING( "ac:hvV" ) ) ) != (libcstring_system_integer_t) -1 )
	{
		switch( option )
		{
			case (libcstring_system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_LIBCSTRING_SYSTEM "\n",
				 argv[ optind - 1 ] );

				usage_fprint(
				 stdout );

				return( EXIT_FAILURE );

			case (libcstring_system_integer_t) 'a':
				show_allocation_information = 1;

				break;

			case (libcstring_system_integer_t) 'c':
				option_ascii_codepage = optarg;

				break;

			case (libcstring_system_integer_t) 'h':
				usage_fprint(
				 stdout );

				return( EXIT_SUCCESS );

			case (libcstring_system_integer_t) 'v':
				verbose = 1;

				break;

			case (libcstring_system_integer_t) 'V':
				pffoutput_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source file.\n" );

		usage_fprint(
		 stdout );

		return( EXIT_FAILURE );
	}
	source = argv[ optind ];

	libcnotify_verbose_set(
	 verbose );
	libpff_notify_set_stream(
	 stderr,
	 NULL );
	libpff_notify_set_verbose(
	 verbose );

	if( info_handle_initialize(
	     &pffinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize info handle.\n" );

		goto on_error;
	}
	if( option_ascii_codepage != NULL )
	{
		result = info_handle_set_ascii_codepage(
		          pffinfo_info_handle,
		          option_ascii_codepage,
		          &error );

		if( result == -1 )
		{
			fprintf(
			 stderr,
			 "Unable to set ASCII codepage in info handle.\n" );

			goto on_error;
		}
		else if( result == 0 )
		{
			fprintf(
			 stderr,
			 "Unsupported ASCII codepage defaulting to: windows-1252.\n" );
		}
	}
/* TODO
	if( libcsystem_signal_attach(
	     pffinfo_signal_handler,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to attach signal handler.\n" );

		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
*/
	if( info_handle_open_input(
	     pffinfo_info_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open: %" PRIs_LIBCSTRING_SYSTEM ".\n",
		 source );

		goto on_error;
	}
	if( info_handle_file_fprint(
	     pffinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to print file information.\n" );

		goto on_error;
	}
	if( info_handle_message_store_fprint(
	     pffinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to print file stream and storage items.\n" );

		goto on_error;
	}
	if( show_allocation_information != 0 )
	{
		if( info_handle_unallocated_blocks_fprint(
		     pffinfo_info_handle,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to print file unallocated blocks.\n" );

			goto on_error;
		}
	}
/* TODO
	if( libcsystem_signal_detach(
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to detach signal handler.\n" );

		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
*/
	if( info_handle_close(
	     pffinfo_info_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close info handle.\n" );

		goto on_error;
	}
	if( info_handle_free(
	     &pffinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free info handle.\n" );

		goto on_error;
	}
	return( EXIT_SUCCESS );

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( pffinfo_info_handle != NULL )
	{
		info_handle_close(
		 pffinfo_info_handle,
		 NULL );
		info_handle_free(
		 &pffinfo_info_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

